producer/buffer/buffer.go

```diff
@@ -0,0 +1,197 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package buffer
+
+import (
+	"container/list"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3msg/producer"
+	"github.com/m3db/m3msg/producer/data"
+
+	"github.com/uber-go/atomic"
+	"github.com/uber-go/tally"
+)
+
+var (
+	errBufferFull                         = errors.New("buffer full")
+	errBufferClosed                       = errors.New("buffer closed")
+	errInvalidDataLargerThanMaxBufferSize = errors.New("invalid data, larger than max buffer size")
+)
+
+type bufferMetrics struct {
+	messageDropped  tally.Counter
+	bytesDropped    tally.Counter
+	messageBuffered tally.Gauge
+	bytesBuffered   tally.Gauge
+}
+
+func newBufferMetrics(scope tally.Scope) *bufferMetrics {
+	return &bufferMetrics{
+		messageDropped:  scope.Counter("message-dropped"),
+		bytesDropped:    scope.Counter("bytes-dropped"),
+		messageBuffered: scope.Gauge("message-buffered"),
+		bytesBuffered:   scope.Gauge("bytes-buffered"),
+	}
+}
+
+type buffer struct {
+	sync.Mutex
+
+	strategy        OnFullStrategy
+	maxBufferSize   uint64
+	cleanupInterval time.Duration
+	buffer          *list.List
+	m               *bufferMetrics
+
+	size     *atomic.Uint64
+	doneCh   chan struct{}
+	isClosed bool
+}
+
+// NewBuffer returns a new buffer.
+func NewBuffer(opts Options) producer.Buffer {
+	if opts == nil {
+		opts = NewBufferOptions()
+	}
+	return &buffer{
+		size:            atomic.NewUint64(0),
+		buffer:          list.New(),
+		isClosed:        false,
+		doneCh:          make(chan struct{}),
+		strategy:        opts.OnFullStrategy(),
+		maxBufferSize:   opts.MaxBufferSize(),
+		cleanupInterval: opts.CleanupInterval(),
+		m:               newBufferMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+}
+
+func (b *buffer) Buffer(d producer.Data) (producer.RefCountedData, error) {
+	b.Lock()
+	if b.isClosed {
+		b.Unlock()
+		return nil, errBufferClosed
+	}
+	var (
+		dataSize      = uint64(d.Size())
+		maxBufferSize = b.maxBufferSize
+	)
+	if dataSize > maxBufferSize {
+		b.Unlock()
+		return nil, errInvalidDataLargerThanMaxBufferSize
+	}
+	targetBufferSize := maxBufferSize - dataSize
+	if b.size.Load() > targetBufferSize {
+		if err := b.produceOnFullWithLock(targetBufferSize); err != nil {
+			b.Unlock()
+			return nil, err
+		}
+	}
+	b.size.Add(dataSize)
+	rd := data.NewRefCountedData(d, b.subSize)
```

Comment:
I think you should store `callback := b.subSize` in the buffer and use `callback` here, otherwise you'll need to allocate a lambda on the heap to capture `b` during every `Buffer` call.